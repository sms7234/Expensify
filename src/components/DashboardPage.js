import React from 'react';
import {connect} from 'react-redux';
import Chart from 'react-google-charts';
import moment from 'moment';
import selectExpenses from '../selectors/expenses'
import DashboardSummary from './DashboardSummary';
import ExpenseListFilters from './ExpenseListFilters';

const DashboardPage = ({income, expenses, pieData, lineData}) => (
  <div>
    <DashboardSummary income={income} expenses={expenses} />
    <div className="content-container">
      <ExpenseListFilters />
      <div className="content-container--charts">
        <h1> Purchases by Category </h1>
        <Chart
          chartType="PieChart"
          height={400}
          loader={<div>Loading Chart</div>}
          data={pieData}
          options={{
            is3D: true,
          }}
          rootProps={{ 'data-testid': '2' }}
        />
        <h1> Remaining Funds (in $) by Date </h1>
        <Chart
          chartType="LineChart"
          height={400}
          loader={<div>Loading Chart</div>}
          data={lineData}
          options={{
            legend: {
              position: "none"
            },
            hAxis: {
              title: 'Remaining Funds',
            },
            vAxis: {
              title: 'Date',
            },
          }}
          rootProps={{ 'data-testid': '1' }}
        />
      </div>
    </div>
  </div>
)

const mapStateToProps=(state)=>{
  //Sum data
  const sumItems = (arr) => {
    let holder = 0;
    if(arr!==[]){
      arr.forEach((item) => {
        holder = holder + item.amount
      })
    }
    return holder/100
  };

  //filter data
  const filterIncome = (arr) => {
    return arr.filter((item) => item.category==='Income')
  };

  const filterExpenses = (arr) => {
    return arr.filter((item)=> item.category!=="Income")
  };

  const filterPie = (arr) => {
    const tempHolder = [];
    const pieData = [['Category', 'Cost']];
    arr.forEach((item) => {
      if(!tempHolder.includes(item.category) && item.category!=="Income") {
        tempHolder.push(item.category)
      }
    });
    tempHolder.forEach((item) => {
      let sum = 0;
      arr.forEach((subItem,index) => {
        if(subItem.category===item) {
          sum = sum + subItem.amount;
        }
      });
      pieData.push([item, sum/100]);
    })
    return pieData;
  }

  const filterLine = (arr) => {
    const lineData=[['Date', 'Remaining Funds']];
    const originalData = arr.sort((a,b) => {
      return a.createdAt < b.createdAt ? -1: 1;
    });
    const len = originalData.length;
    let total=0;

    for (let i=0; i<len; i++){
      const firstDate = moment(originalData[i].createdAt).format("MM/DD/YY");
      if(i<(len-1)) { //be sure i isn't last item
        for(let j=1;j<=len;j++){
          if(j+i<len){ //be sure not to overrun the array length
            const secondDate = moment(originalData[i+j].createdAt).format("MM/DD/YY");
            if(firstDate === secondDate) {
              if(originalData[i].category === 'Income'){
                total = total + originalData[i].amount
              } else {
                total = total - originalData[i].amount
              }
            } else {
              lineData.push([firstDate, total/100])
              i=i+(j-1);
              break;
            }
          } else {
            break;
          }
        }
      } else { //if last item in array push info to array
        if(originalData[i].category === 'Income'){
          total = total + originalData[i].amount
        } else {
          total = total - originalData[i].amount
        }
        lineData.push([firstDate,total/100])
      }
    }
    return lineData;
  }
  // actual data
    let visibleExpenses = selectExpenses(state.expenses, state.filters);
    let selectedExpenses = sumItems(filterExpenses(visibleExpenses));

  return {
    a: selectExpenses(state.expenses, state.filters),
    expenses: sumItems(filterExpenses(visibleExpenses)),
    income: sumItems(filterIncome(visibleExpenses)),
    pieData: filterPie(selectExpenses(state.expenses, state.filters)),
    lineData: filterLine(selectExpenses(state.expenses, state.filters))
  };
};

export default connect(mapStateToProps)(DashboardPage);
