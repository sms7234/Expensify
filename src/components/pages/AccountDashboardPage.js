import React from 'react';
import AccountList from '../lists/AccountList';
import AccountSummary from '../independents/AccountSummary';
import AccountListFilters from '../filters/AccountListFilters';

const AccountDashboardPage = () => (
  <div>
    <AccountSummary />
    <AccountListFilters />
    <AccountList />
  </div>
)

export default AccountDashboardPage;
