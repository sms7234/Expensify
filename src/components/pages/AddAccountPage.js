import React from 'react';
import AccountForm from '../forms/AccountForm';
import {connect} from 'react-redux';
import {startAddAccount} from '../../actions/accounts';

export class AddAccountPage extends React.Component{
  onSubmit=(account) => {
    this.props.startAddAccount(account);
    this.props.history.push('/accounts');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title"> Add Account </h1>
          </div>
        </div>
        <div className="content-container">
          <AccountForm onSubmit={this.onSubmit} />
        </div>
      </div>
    );
  };
};

const mapDispatchToProps = (dispatch) => ({
  startAddAccount:(account) => dispatch(startAddAccount(account))
});

export default connect(undefined, mapDispatchToProps)(AddAccountPage);
