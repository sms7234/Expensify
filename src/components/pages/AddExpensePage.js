import React from 'react';
import ExpenseForm from '../forms/ExpenseForm';
import { connect} from 'react-redux';
import {startAddExpense} from '../../actions/expenses';

export class AddExpensePage extends React.Component{
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/expenses');
  };
  render(){
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title"> Add Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            onSubmit = {this.onSubmit}
            categoryList = {this.props.categoryList}
            accountList = {this.props.accountList}
            tagList = {this.props.tagList}
          />
        </div>
      </div>
    );
  }
};

const mapStateToProps=(state) => {
  return {
    accountList: state.accounts.map((item)=>{return({label: item.account, value: item.account})}),
    categoryList: state.categories.map((item)=>{return({label: item.category, value: item.category})}),
    tagList: state.tags.map((item)=>{return({label: item.tag, value: item.tag})})
  };
};

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(mapStateToProps, mapDispatchToProps)(AddExpensePage);
