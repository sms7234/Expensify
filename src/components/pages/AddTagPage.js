import React from 'react';
import TagForm from '../forms/TagForm';
import {connect} from 'react-redux';
import {startAddTag} from '../../actions/tags';

export class AddTagPage extends React.Component{
  onSubmit=(tag) => {
    this.props.startAddTag(tag);
    this.props.history.push('/tags');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title"> Add Tag </h1>
          </div>
        </div>
        <div className="content-container">
          <TagForm onSubmit={this.onSubmit} />
        </div>
      </div>
    );
  };
};

const mapDispatchToProps = (dispatch) => ({
  startAddTag:(tag) => dispatch(startAddTag(tag))
});

export default connect(undefined, mapDispatchToProps)(AddTagPage);
