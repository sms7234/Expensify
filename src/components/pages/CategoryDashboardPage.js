import React from 'react';
import CategoryList from '../lists/CategoryList';
import CategoryListFilters from '../filters/CategoryListFilters';
import CategorySummary from '../independents/CategorySummary';

const CategoryDashboardPage = () => (
  <div>
    <CategorySummary />
    <CategoryListFilters />
    <CategoryList />
  </div>
)

export default CategoryDashboardPage;
