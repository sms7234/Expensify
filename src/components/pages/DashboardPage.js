import React from 'react';
import {connect} from 'react-redux';
import {Doughnut, Line} from 'react-chartjs-2';
import moment from 'moment';
import selectExpenses from '../../selectors/expenses'
import DashboardSummary from '../independents/DashboardSummary';
import ExpenseListFilters from '../filters/ExpenseListFilters';

export const DashboardPage = ({income, expenses, pieData, lineData, donutData}) => (
  <div>
    <DashboardSummary income={income} expenses={expenses} />
    <div className="content-container">
      <ExpenseListFilters />
      <div className="content-container--charts">
        <h1> Purchases by Category </h1>
        <Doughnut
          data={donutData}
        />
        <span />
        <h1> Remaining Funds (in $) by Date </h1>
        <Line
          data={lineData}
        />
      </div>
    </div>
  </div>
)

const mapStateToProps=(state)=>{
  //color array
  const colors = ['#E74C3C','#9B59B6', '#2980B9', '#1ABC9C', '#27AE60', '#F1C40F', '#F39C12', '#D35400', '#C0392B', '#8E44AD', '#3498DB', '#16A085', '#2ECC71', '#F5B041', '#EB984E' ];

  //assign colors - pass in array with desired length
  const colorAssign = (desiredQty) => {
    const max = colors.length;
    const holder = [];
    let n=0;
    desiredQty.forEach((item,index) => {
      if (index === max*(n+1)){n = n+1}
      holder.push(colors[index-(n*max)])
    })
    return holder;
  }

  //Sum data
  const sumItems = (arr) => {
    let holder = 0;
    if(arr!==[]){
      arr.forEach((item) => {
        holder = holder + item.amount
      })
    }
    return holder/100
  };

  //filter data
  const filterIncome = (arr) => {
    return arr.filter((item) => item.category==='Income')
  };

  const filterExpenses = (arr) => {
    return arr.filter((item)=> item.category!=="Income")
  };

  const FilterDoughnut = (arr) => {
    const labelHolder = {};

    //generate labels
    arr.forEach((item) => {
      if(!labelHolder.hasOwnProperty(item.category) && item.category!=="Income") {
        labelHolder[item.category]=0;
      }
    });
    //sum up values for each label
    arr.forEach((item) => {
      if(item.category !== "Income") {
        labelHolder[item.category] += item.amount/100;
      }
    })
    //generate colors
    const colorHolder = colorAssign(Object.keys(labelHolder));

    //reformat & returndata
    const donutData = {datasets:[{data: Object.values(labelHolder), backgroundColor: colorHolder}], labels:Object.keys(labelHolder)}
    return donutData;
  };

  const filterLineData = (arr) => {
    //holders
    const accounts = {};
    const labels= [];
    let counter = 0
    //sort by date
    const originalData = arr.sort((a,b) => {
      return a.purchaseDate < b.purchaseDate ? -1: 1;
    });
    //pull out unique dates & accounts
    originalData.forEach((item) => {
      const curDate = moment(item.purchaseDate).format('MM/DD/YY');
      if (!labels.includes(curDate)){
        labels.push(curDate)
      }
      if(!accounts.hasOwnProperty(item.account)) {
        accounts[item.account]=[0];
      }
    })
    //run through array & sort out expenses into acctObj
    let lastMatch = 0;
    const len = originalData.length;
    const accountList = Object.keys(accounts);
    labels.forEach((item,index) => {
      //add prev date sum (in accounts array) as the next item in array
      if(index>0){
        accountList.forEach((item) => {
          accounts[item].push(accounts[item][index-1]);
        })
      }
      for(let i=lastMatch; i<len; i++){
        //dates are the same...
        if(moment(item).isSame(originalData[i].purchaseDate, 'day')) {
          const acctName = originalData[i].account;
          if(originalData[i].category === 'Income'){
            accounts[acctName][index] += originalData[i].amount/100
          } else {
            accounts[acctName][index] -= originalData[i].amount/100
          }
          lastMatch = i;
          counter +=1;
        }
        //item date is greater than label date
        if(moment(originalData[i].purchaseDate).isAfter(item, 'day')) {
          break;
        }
      }
    })

    //assign colors
    const colorHolder = colorAssign(accountList);

    //reformat data for chart
    const datasets = [];
    accountList.forEach((item, index) => datasets.push({label: item, data: accounts[item], backgroundColor: colorHolder[index], borderColor: colorHolder[index], fill: false}))
    return {labels, datasets}

    //return correctly formatted information
    console.log('labels: ',labels, 'datasets: ',datasets)
    console.log('number of matches: ', counter, 'total expenses: ', originalData.length)

  }

  // actual data
    let visibleExpenses = selectExpenses(state.expenses, state.filters);
    let selectedExpenses = sumItems(filterExpenses(visibleExpenses));

  return {
    expenses: sumItems(filterExpenses(visibleExpenses)),
    income: sumItems(filterIncome(visibleExpenses)),
    lineData: filterLineData(selectExpenses(state.expenses, state.filters)),
    donutData: FilterDoughnut(selectExpenses(state.expenses, state.filters))
  };
};

export default connect(mapStateToProps)(DashboardPage);
