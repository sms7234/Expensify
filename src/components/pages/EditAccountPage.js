import React from 'react';
import {connect} from 'react-redux';
import AccountForm from '../forms/AccountForm';
import {startEditAccount, startRemoveAccount} from '../../actions/accounts';

export class EditAccountPage extends React.Component{
  onSubmit=(account)=>{
    this.props.startEditAccount(this.props.accounts.id, account),
    this.props.history.push('/accounts')
  };
  onRemove= ()=> {
    this.props.startRemoveAccount({id: this.props.accounts.id}),
    this.props.history.push('/accounts')
  };
  render(){
    return(
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title"> Edit Account</h1>
          </div>
        </div>
        <div className="content-container">
          <AccountForm
            accounts={this.props.accounts}
            onSubmit={this.onSubmit}
          />
          <button className="button--secondary" onClick={this.onRemove}>Remove Account</button>
        </div>
      </div>
    )
  }
};

const mapStateToProps = (state, props) => {
  return {
    accounts: state.accounts.find((item) => item.id === props.match.params.id)
  };
};

const mapDispatchToProps = (dispatch, props) => ({
  startEditAccount: (id,account) => dispatch(startEditAccount(id,account)),
  startRemoveAccount: (data) => dispatch(startRemoveAccount(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditAccountPage);
