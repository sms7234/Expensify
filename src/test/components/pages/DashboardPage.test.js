import React from 'react';
import {shallow} from 'enzyme';
import {DashboardPage} from '../../../components/pages/DashboardPage';

const fakeData = {
  income: 2400,
  expenses: 1500,
  lineData: {labels:["04/01/20", "04/02/20", "04/03/20"], datasets:[{label: "BoA", data: [150,200,300]}, {label: "Chase", data: [300,200,100]}]},
  donutData: {labels:['Groceries', 'Eating Out'], datasets: [{label:'a', data:[500,100]}]},
  monthData:{labels:["03/20", "04/20"], datasets:[{label:"expenses", data:[500,700]}, {label:"income", data: [1000, 1000]}]},
  categoryData:{'03/20':{"Groceries": 500, "Eating Out":200}, '04/20': {"Groceries": 1000, "Eating Out": 100}},
  categoryList: [{label: "Groceries", value: "Groceries"}, {label: "Eating Out", value: "Eating Out"}]
};

test('should render page with no data', () => {
  const wrapper = shallow(<DashboardPage />);
  expect(wrapper).toMatchSnapshot();
});

test('should render page with fake data', () => {
  const wrapper = shallow(<DashboardPage income={fakeData.income} expenses={fakeData.expenses} donutData={fakeData.donutData} lineData={fakeData.lineData} monthData={fakeData.monthData} categoryList={fakeData.categoryList} categoryData={fakeData.categoryData}/>)
  expect(wrapper).toMatchSnapshot();
})
