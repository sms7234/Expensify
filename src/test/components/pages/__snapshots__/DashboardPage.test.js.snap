// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render page with fake data 1`] = `
<div>
  <DashboardSummary
    expenses={1500}
    income={2400}
  />
  <div
    className="content-container"
  >
    <Connect(ExpenseListFilters) />
    <div
      className="content-container"
    >
      <button
        className="button--toggle"
        onClick={[Function]}
      >
         Toggle Legends
      </button>
      <Tabs
        className="tabs"
        defaultActiveKey="pie"
        id="uncontrolled-tab-example"
        mountOnEnter={false}
        unmountOnExit={false}
        variant="tabs"
      >
        <Tab
          eventKey="pie"
          title="Expenses Breakdown"
        >
          <div
            className="content-container--charts"
          >
            <h1>
               Purchases by Category 
            </h1>
            <Doughnut
              data={
                Object {
                  "datasets": Array [
                    Object {
                      "data": Array [
                        500,
                        100,
                      ],
                      "label": "a",
                    },
                  ],
                  "labels": Array [
                    "Groceries",
                    "Eating Out",
                  ],
                }
              }
              options={
                Object {
                  "legend": Object {
                    "display": false,
                    "position": "right",
                  },
                }
              }
            />
          </div>
        </Tab>
        <Tab
          eventKey="line"
          title="Account Breakdown"
        >
          <div
            className="content-container--charts"
          >
            <h1>
               Remaining Funds (in $) by Date 
            </h1>
            <Line
              data={
                Object {
                  "datasets": Array [
                    Object {
                      "data": Array [
                        150,
                        200,
                        300,
                      ],
                      "label": "BoA",
                    },
                    Object {
                      "data": Array [
                        300,
                        200,
                        100,
                      ],
                      "label": "Chase",
                    },
                  ],
                  "labels": Array [
                    "04/01/20",
                    "04/02/20",
                    "04/03/20",
                  ],
                }
              }
              options={
                Object {
                  "legend": Object {
                    "display": false,
                  },
                }
              }
            />
          </div>
        </Tab>
        <Tab
          eventKey="bar"
          title="Income vs Expense"
        >
          <div
            className="content-container--charts"
          >
            <h1>
               Monthly Summary 
            </h1>
            <Bar
              data={
                Object {
                  "datasets": Array [
                    Object {
                      "data": Array [
                        500,
                        700,
                      ],
                      "label": "expenses",
                    },
                    Object {
                      "data": Array [
                        1000,
                        1000,
                      ],
                      "label": "income",
                    },
                  ],
                  "labels": Array [
                    "03/20",
                    "04/20",
                  ],
                }
              }
              options={
                Object {
                  "legend": Object {
                    "display": false,
                  },
                }
              }
            />
          </div>
        </Tab>
        <Tab
          eventKey="variable-bar"
          title="Category History"
        >
          <div
            className="content-container--charts"
          >
            <h1>
               Category History 
            </h1>
            <h3>
              add/remove desired categories
            </h3>
            <p>
              ***Warning*** Information only updates when categories are change
            </p>
            <p>
              There is no auto-update when dates change
            </p>
            <StateManager
              className="dropdown"
              defaultInputValue=""
              defaultMenuIsOpen={false}
              defaultValue={null}
              isClearable={true}
              isMulti={true}
              onChange={[Function]}
              options={
                Array [
                  Object {
                    "label": "Groceries",
                    "value": "Groceries",
                  },
                  Object {
                    "label": "Eating Out",
                    "value": "Eating Out",
                  },
                ]
              }
              placeholder="Select Categories to show"
            />
            <p>
              Chart HERE 
            </p>
            <Bar
              data={
                Object {
                  "datasets": Array [
                    Object {
                      "data": Array [
                        0,
                      ],
                    },
                  ],
                  "labels": Array [
                    "no data",
                  ],
                }
              }
              options={
                Object {
                  "legend": Object {
                    "display": false,
                  },
                }
              }
            />
          </div>
        </Tab>
      </Tabs>
    </div>
  </div>
</div>
`;

exports[`should render page with no data 1`] = `
<div>
  <DashboardSummary />
  <div
    className="content-container"
  >
    <Connect(ExpenseListFilters) />
    <div
      className="content-container"
    >
      <button
        className="button--toggle"
        onClick={[Function]}
      >
         Toggle Legends
      </button>
      <Tabs
        className="tabs"
        defaultActiveKey="pie"
        id="uncontrolled-tab-example"
        mountOnEnter={false}
        unmountOnExit={false}
        variant="tabs"
      >
        <Tab
          eventKey="pie"
          title="Expenses Breakdown"
        >
          <div
            className="content-container--charts"
          >
            <h1>
               Purchases by Category 
            </h1>
            <Doughnut
              options={
                Object {
                  "legend": Object {
                    "display": false,
                    "position": "right",
                  },
                }
              }
            />
          </div>
        </Tab>
        <Tab
          eventKey="line"
          title="Account Breakdown"
        >
          <div
            className="content-container--charts"
          >
            <h1>
               Remaining Funds (in $) by Date 
            </h1>
            <Line
              options={
                Object {
                  "legend": Object {
                    "display": false,
                  },
                }
              }
            />
          </div>
        </Tab>
        <Tab
          eventKey="bar"
          title="Income vs Expense"
        >
          <div
            className="content-container--charts"
          >
            <h1>
               Monthly Summary 
            </h1>
            <Bar
              options={
                Object {
                  "legend": Object {
                    "display": false,
                  },
                }
              }
            />
          </div>
        </Tab>
        <Tab
          eventKey="variable-bar"
          title="Category History"
        >
          <div
            className="content-container--charts"
          >
            <h1>
               Category History 
            </h1>
            <h3>
              add/remove desired categories
            </h3>
            <p>
              ***Warning*** Information only updates when categories are change
            </p>
            <p>
              There is no auto-update when dates change
            </p>
            <StateManager
              className="dropdown"
              defaultInputValue=""
              defaultMenuIsOpen={false}
              defaultValue={null}
              isClearable={true}
              isMulti={true}
              onChange={[Function]}
              placeholder="Select Categories to show"
            />
            <p>
              Chart HERE 
            </p>
            <Bar
              data={
                Object {
                  "datasets": Array [
                    Object {
                      "data": Array [
                        0,
                      ],
                    },
                  ],
                  "labels": Array [
                    "no data",
                  ],
                }
              }
              options={
                Object {
                  "legend": Object {
                    "display": false,
                  },
                }
              }
            />
          </div>
        </Tab>
      </Tabs>
    </div>
  </div>
</div>
`;
